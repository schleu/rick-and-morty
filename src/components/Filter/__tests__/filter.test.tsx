import { fireEvent, render, screen } from '@testing-library/react';
import { Filter, FilterProps } from '../index';
import userEvent from '@testing-library/user-event';

import '@testing-library/jest-dom'

const mockSetFilters = jest.fn();

const mockFilter: FilterProps = {
  name: 'Rick',
  gender: 'Male',
  status: 'Alive',
};

describe('<Filter />', () => {
  it('snapshot', () => {
    const { container } = render(<Filter />);
    expect(container).toMatchSnapshot()
  });

  it('should render properly', () => {
    render(<Filter />);

    const nameInput = screen.getByTestId('name') as HTMLInputElement;
    const genderSelect = screen.getByTestId('gender') as HTMLSelectElement;
    const statusSelect = screen.getByTestId('status') as HTMLSelectElement;
    const searchButton = screen.getByTestId('search') as HTMLButtonElement;

    expect(nameInput).toBeInTheDocument();
    expect(genderSelect).toBeInTheDocument();
    expect(statusSelect).toBeInTheDocument();
    expect(searchButton).toBeInTheDocument();
  });

  it('should be able change values', async () => {
    const { getByTestId } = render(<Filter />);

    const nameInput = getByTestId('name') as HTMLInputElement;
    const genderSelect = getByTestId('gender') as HTMLSelectElement;
    const statusSelect = getByTestId('status') as HTMLSelectElement;

    await userEvent.type(nameInput, String(mockFilter.name));
    fireEvent.change(genderSelect, { target: { value: mockFilter.gender} });
    fireEvent.change(statusSelect, { target: { value: mockFilter.status } });

    expect(nameInput).toHaveValue(mockFilter.name)
    expect(genderSelect).toHaveValue(mockFilter.gender)
    expect(statusSelect).toHaveValue(mockFilter.status)

  })

  it('should call setFilters with correct values', async () => {

    const { getByTestId } = render(<Filter />);

    const nameInput = getByTestId('name') as HTMLInputElement;
    const genderSelect = getByTestId('gender') as HTMLSelectElement;
    const statusSelect = getByTestId('status') as HTMLSelectElement;
    const searchButton = getByTestId('search') as HTMLButtonElement;

    await userEvent.type(nameInput, String(mockFilter.name));
    fireEvent.change(genderSelect, { target: { value: mockFilter.gender} });
    fireEvent.change(statusSelect, { target: { value: mockFilter.status } });
    await userEvent.click(searchButton)

    expect(mockSetFilters).toHaveBeenCalledWith(mockFilter);
  });
});
