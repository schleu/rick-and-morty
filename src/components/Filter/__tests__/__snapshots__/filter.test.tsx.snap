// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Filter /> snapshot 1`] = `
<div>
  <div
    class="filter"
  >
    <div
      class="title"
    >
      Filtro
    </div>
    <form
      data-testid="filterForm"
    >
      <input
        data-testid="name"
        name="name"
        placeholder="personagem"
        type="text"
      />
      <select
        data-testid="gender"
        name="gender"
      >
        <option
          value=""
        >
          Gender
        </option>
        <option
          value="Female"
        >
          Female
        </option>
        <option
          value="Male"
        >
          Male
        </option>
        <option
          value="Genderless"
        >
          Genderless
        </option>
        <option
          value="unknown"
        >
          unknown
        </option>
      </select>
      <select
        data-testid="status"
        name="status"
      >
        <option
          value=""
        >
          Status
        </option>
        <option
          value="Alive"
        >
          Alive
        </option>
        <option
          value="Dead"
        >
          Dead
        </option>
        <option
          value="unknown"
        >
          unknown
        </option>
      </select>
      <button
        data-testid="search"
        type="submit"
      >
        Pesquisar
      </button>
    </form>
  </div>
</div>
`;
